### ISSUE TITLE
Robustness of literal math_equal()

### ISSUE THREAD

**Issue by rupertford** [stfc/PSyclone#1321]
math_equal() in a Literal is not robust. It does a direct string comparison and does not take into account datatype - although it is unclear to me what is correct in the latter case.

For example:
Literal("1.0", REAL_TYPE) and Literal("1.00", REAL_TYPE) are flagged as not being equal
Literal("1.0", REAL_TYPE) and Literal("1.0", REAL_SINGLE_TYPE) are flagged as being the same (is this right?)
Literal("1", REAL_TYPE) and Literal("1" INTEGER) are flagged as the same (is this right?)

**Comment by hiker** [comment]
Agreed, it only compares the `self._value` content - whatever that might be :) Which explains the first failure :)

I am pretty certain that numbers in a comparison are extended to the larger type. So int vs real means the int is converted to real. So from that point of view the other two comparisons would be correct (I would assume that comparing real 1.0 with int 1 would still return wrong with the current code).

I would suggest to postpone this till we have support for symbolic math, which should handle this case automatically, see #1075 (well, it's part of that issue). We need to make a decision if we want to require sympy, or make it only optional. If the latter, then we would indeed need to fix math_equal (in whatever form it will be refactored as part of #1075).
